RFC-0016-Masked-reductions-and-normalizations.md

```diff
@@ -0,0 +1,259 @@
+
+
+
+* There is some explicit user demand for reductions with masked semantics for sparse Tensors
+    * [[feature request] Sparse (hybrid sparse-dense) output option for topk, min, max ](https://github.com/pytorch/pytorch/issues/34951)
+    * [Feature request: sparse matrix max(axis)](https://github.com/pytorch/pytorch/issues/4247) 
+    * [Implemented torch.sparse.max(sparse_tensor, dim)](https://github.com/pytorch/pytorch/pull/59340)
+* [RFC-0004: Adding fill value property to PyTorch sparse tensors](https://github.com/pytorch/rfcs/pull/8) is blocked on discussions around masked semantics.
+* There are also demands for masked normalizations for dense Tensors and some prior art that shows performance improvements.
+    * [Add masked_softmax to speed up masking in multihead attention](https://github.com/pytorch/pytorch/pull/48441)
+    * [Fuse softmax and masking in MultiheadAttention](https://github.com/pytorch/pytorch/issues/44945)
+    * [[FR] Safe softmax](https://github.com/pytorch/pytorch/issues/55056)   
+
+**Existing reduction operators**
+
+Reductions are operators that remove a dimension from a Tensor. They apply a binary operation across a list of Tensor slices and aggregate the result. This is different from normalizations that return the same shape with all dimensions intact.
+
+Dense ([docs](https://pytorch.org/docs/master/torch.html#reduction-ops))
+
+
+
+* argmax
+* argmin
+* amax
+* amin
+* aminmax
+* all
+* any
+* max
+* min
+* dist
+* logsumexp
+* mean
+* median
+* nanmedian
+* mode
+* norm
+* nansum
+* prod
+* quantile
+* nanquantile
+* std
+* std_mean
+* sum
+* unique
+* unique_consecutive
+* var
+* var_mean
+* count_nonzero
+
+Sparse COO ([docs](https://pytorch.org/docs/master/sparse.html#torch-functions-specific-to-sparse-tensors))
+
+
+
+* sum
+
+Sparse CSR: None
+
+**Existing normalization operators**
+
+These are a bit more difficult to list exhaustively, because we don't have an explicit category for them. A normalization operator applies a reduction to a Tensor slice and uses the result to normalize the slice.
+
+Dense ([functionals](https://pytorch.org/docs/master/nn.functional.html))
+
+
+
+* softmax
+* log_softmax
+* batch_norm
+* group_norm
+* layer_norm
+* instance_norm
+* local_response_norm
+* normalize
+* cross_entropy
+* nll_loss
+
+Sparse COO ([docs](https://pytorch.org/docs/master/sparse.html#torch-functions-specific-to-sparse-tensors))
+
+
+
+* softmax
+* log_softmax
+* norm / native_norm
+
+Sparse CSR: None
+
+**Operator constraints and general signature - Reductions**
+
+**Input types:** A mask is a boolean tensor. It accompanies a dense Tensor of the same shape. If an entry is True the corresponding element at the same index in the paired dense Tensor is a "valid" value. If it is False it is not. "valid" here means that this value is meant to be included in the computation and otherwise is meant to be ignored. This matches the semantics of masked_scatter, masked_select and masked_fill. 
```

Comment:
This definition of mask mismatches the definition of [numpy.ma mask](https://numpy.org/doc/stable/reference/maskedarray.generic.html#what-is-a-masked-array):
```
When an element of the mask is False, the corresponding element of the associated array is valid
and is said to be unmasked. When an element of the mask is True, the corresponding element of
the associated array is said to be masked (invalid).
```
Possible solutions:
1. Adjust the mask definition to match the one in numpy.ma
2. Rename "mask" to "valid"
3. Document the mismatch with numpy.ma

I would vote for 1 because the mismatch with the three non-arithmetic ops `masked_scatter`, ... would not be as bad as the mismatch with arithmetic ops of more widely used numpy.ma module, IMHO.

Reply:
I will redraw my vote for option 1 and suggest defining:
```
A mask is a boolean tensor of selection.
```

Reply:
I think the current definition of mask adheres to ```A mask is a boolean tensor of selection.``` if I understand you correctly?

Interestingly enough our [MHA implementation](https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention) uses True to indicate an element is meant to be ignored.

But like you mention the other ops use the inverse of this definition.

Reply:
> I think the current definition of mask adheres to `A mask is a boolean tensor of selection.` if I understand you correctly?

yes

> Interestingly enough our [MHA implementation](https://pytorch.org/docs/master/generated/torch.nn.MultiheadAttention.html#torch.nn.MultiheadAttention) uses True to indicate an element is meant to be ignored.

I guess it boils down to not assuming that "mask" is a uniquely defined concept and the exact meaning of mask must be documented in functions docstrings.

> But like you mention the other ops use the inverse of this definition.

In addition, https://numpy.org/doc/stable/reference/generated/numpy.sum.html uses `where` argument as a "mask of selection".

Reply:
Based on our offline discussion with Ralf I think it's safe to go for mask semantics that adhere to those described in this RFC.